pkg/controller/persistentvolume/index.go

```diff
@@ -75,27 +74,38 @@ func (pvIndex *persistentVolumeOrderedIndex) findByClaim(claim *api.PersistentVo
 	// potential matches (the GCEPD example above).
 	allPossibleModes := pvIndex.allPossibleMatchingAccessModes(claim.Spec.AccessModes)
 
+	var smallestVolume *api.PersistentVolume
+	var smallestVolumeSize int64
+	requestedQty := claim.Spec.Resources.Requests[api.ResourceName(api.ResourceStorage)]
+	requestedSize := requestedQty.Value()
+
 	for _, modes := range allPossibleModes {
 		volumes, err := pvIndex.listByAccessModes(modes)
 		if err != nil {
 			return nil, err
 		}
 
-		// volumes are sorted by size but some may be bound or earmarked for a specific claim.
-		// filter those volumes for easy binary search by size
-		// return the exact pre-binding match, if found
-		unboundVolumes := []*api.PersistentVolume{}
+		// Go through all available volumes and find the smallest matching one.
+		// Return the exact pre-binding match, if found.
 		for _, volume := range volumes {
-			// volume isn't currently bound or pre-bound.
-			if volume.Spec.ClaimRef == nil {
-				unboundVolumes = append(unboundVolumes, volume)
-				continue
-			}
-
 			if isVolumeBoundToClaim(volume, claim) {
 				// Exact match! No search required.
 				return volume, nil
 			}
+
+			if volume.Spec.ClaimRef != nil {
+				// This volume waits for exact claim or is alredy bound
+				continue
+			}
+
+			volumeQty := volume.Spec.Capacity[api.ResourceStorage]
+			volumeSize := volumeQty.Value()
+			if volumeSize >= requestedSize {
```

Comment:
I do not agree here. We must loop through all volumes to make sure there is none that is "bound-by-user" or dynamically provisioned to the claim, these have precedence. See the `Exact match! No search required.` few lines above. Volumes dynamically provisioned by the controller are matched there.

I added some comments around these parts.